Export Data Solution

blaggacao opened the pull request:
Due to my limited technical knowledge, I want to initiate discussion about best strategies in order to implement:
- proper relational fields export (eval or ref?)
- xml cleanup
- solve the "ordering on import" problem

I would kindly ask for inline comments on the ideas laid out in the diffs, so we can evolve some code..
ping @kurkop

Commit:
[ADD] Initial version of prototype

Commit:
[ADD] Initial version of prototype

Commit:
[IMP] Add wizard for API version and templates. Update translation file

Commit:
[IMP] Add wizard for API version and templates. Update translation file

Commit:
[ADD] Security template files

Commit:
[ADD] Security template files

Commit:
[ADD] Filters for data and demo data

Commit:
[ADD] Filters for data and demo data

Commit:
[IMP] prototype: implementation of jinja syntax.

Commit:
[IMP] prototype: implementation of jinja syntax.

Commit:
[IMP] prototype: Switched to v8 api.

Commit:
[IMP] prototype: Switched to v8 api.

Commit:
[IMP] prototype: renamed the templates to avoid confusion with pep8 tests.

Commit:
[IMP] prototype: renamed the templates to avoid confusion with pep8 tests.

Commit:
[IMP] prototype: renamed some fields in prototype.py.

Commit:
[IMP] prototype: renamed some fields in prototype.py.

Commit:
[IMP] prototype: implemented the process to create the zip file to export.

Commit:
[IMP] prototype: implemented the process to create the zip file to export.

Commit:
[IMP] prototype: worked on the generation of __openerp__.py and __init__.py files from the prototype.

Commit:
[IMP] prototype: worked on the generation of __openerp__.py and __init__.py files from the prototype.

Commit:
[IMP] prototype: pep8

Commit:
[IMP] prototype: pep8

Commit:
[IMP] prototype: added more comments.

Commit:
[IMP] prototype: added more comments.

Commit:
[IMP] prototype: worked on fields.

Commit:
[IMP] prototype: worked on fields.

Commit:
[IMP] prototype: implemented the generation of files from models/model_name.py.template and models/init.py.template

Commit:
[IMP] prototype: implemented the generation of files from models/model_name.py.template and models/init.py.template

Commit:
[IMP] prototype: first work on implementation of generations of views.

Commit:
[IMP] prototype: first work on implementation of generations of views.

Commit:
[IMP] prototype: improved the arch section in model_view.xml.template

Commit:
[IMP] prototype: improved the arch section in model_view.xml.template

Commit:
[IMP] prototype: Reviewed the names of many2many from plurials to name_ids.

Commit:
[IMP] prototype: Reviewed the names of many2many from plurials to name_ids.

Commit:
[IMP] prototype: added menus.

Commit:
[IMP] prototype: added menus.

Commit:
[IMP] prototype: typo in templates/8.0/views/model_view.xml.template

Commit:
[IMP] prototype: typo in templates/8.0/views/model_view.xml.template

Commit:
[IMP] prototype: improved template with adding extension to header.template to ease management of file with licence header.

Commit:
[IMP] prototype: improved template with adding extension to header.template to ease management of file with licence header.

Commit:
[IMP] prototype: prototype.prototype demo data

Commit:
[IMP] prototype: prototype.prototype demo data

Commit:
[IMP] prototype: worked to get the views and menus properly generated. Added actions to model_menus.xml.template.

Commit:
[IMP] prototype: worked to get the views and menus properly generated. Added actions to model_menus.xml.template.

Commit:
[IMP] prototype: should fix pep8 in model_name.py.template.

Commit:
[IMP] prototype: should fix pep8 in model_name.py.template.

blaggacao committed:
Inline comments :: 'js':[],'css':[],'qweb':[]

I propose to add inline comments for the following reason:

General understanding is promoted when reading the source files as probablyt the main entry point for a developer.
js,css,qweb are taken from other manifests I've seen. They are explained in the link. I think completedness of the manifest scaffold also helps reduce ambiguity about it.

Lastly, I introduced newlines for better readability and grouping.

I hope these changes, well (marginally) improve the usefulness of this great module.

blaggacao committed:
Inline comments :: 'js':[],'css':[],'qweb':[]

I propose to add inline comments for the following reason:

General understanding is promoted when reading the source files as probablyt the main entry point for a developer.
js,css,qweb are taken from other manifests I've seen. They are explained in the link. I think completedness of the manifest scaffold also helps reduce ambiguity about it.

Lastly, I introduced newlines for better readability and grouping.

I hope these changes, well (marginally) improve the usefulness of this great module.

Commit:
[IMP] Module description and dependencies

Commit:
[IMP] Module description and dependencies

Commit:
[FIX] URL

Commit:
[FIX] URL

Commit:
[FIX] URL to issue

Commit:
[FIX] URL to issue

Commit:
[FIX] Typo

Commit:
[FIX] Typo

Commit:
[FIX] Typo

Commit:
[FIX] Typo

Commit:
[IMP] prototype: added helper fields and placeholders on new fields.

Commit:
[IMP] prototype: added helper fields and placeholders on new fields.

Commit:
[IMP] prototype: no more bug when no icon, commented no implemented page in the view.

Commit:
[IMP] prototype: no more bug when no icon, commented no implemented page in the view.

Commit:
[IMP] prototype: better management of special cases.

Commit:
[IMP] prototype: better management of special cases.

Commit:
[IMP] prototype: prototype.py removed print statement and shadowing of "fields".

Commit:
[IMP] prototype: prototype.py removed print statement and shadowing of "fields".

Commit:
[IMP] prototype -> module_prototyper

Commit:
[IMP] prototype -> module_prototyper

Commit:
[IMP] module_prototyper: gathered menuitems that is helpfull to create a prototype under the menu 'Module Prototypes'

Commit:
[IMP] module_prototyper: gathered menuitems that is helpfull to create a prototype under the menu 'Module Prototypes'

Commit:
[IMP] module_prototyper: more comment and docstrings.

Commit:
[IMP] module_prototyper: more comment and docstrings.

Commit:
[IMP] module_prototyper: translations.

Commit:
[IMP] module_prototyper: translations.

Commit:
[FIX] module_prototyper: pep8

Commit:
[FIX] module_prototyper: pep8

Commit:
[FIX] module_prototyper: fix typo in tests.

Commit:
[FIX] module_prototyper: fix typo in tests.

Commit:
[IMP] update module_prototyper:

* pump up the version of the module to 0.3
* replace ir.ui.model by ir.ui.view in generated xml views
* improve pep8 compatibility of generation of models

Commit:
[IMP] update module_prototyper:

* pump up the version of the module to 0.3
* replace ir.ui.model by ir.ui.view in generated xml views
* improve pep8 compatibility of generation of models

Commit:
[IMP] update module_prototyper: update of README with versions.

Commit:
[IMP] update module_prototyper: update of README with versions.

Commit:
[IMP] Add help

Commit:
[IMP] Add help

Commit:
[FIX] license field
[IMP] Add template README.rst as default description

Commit:
[FIX] license field
[IMP] Add template README.rst as default description

Commit:
[FIX] PEP8 errors

Commit:
[FIX] PEP8 errors

Commit:
[IMP] Take reviews into account

Commit:
[IMP] Take reviews into account

Commit:
[IMP] module_prototyper: add placeholders to fields.

Commit:
[IMP] module_prototyper: add placeholders to fields.

Commit:
[FIX] E402 module level import not at top of file

Commit:
[FIX] E402 module level import not at top of file

ehdem committed:
[UPD] Add a new file for default value decription field

ehdem committed:
[UPD] Add a new file for default value decription field

Commit:
fix template description being indented

Commit:
fix template description being indented

Commit:
[FIX] Use human name and english category name for manifest

Commit:
[FIX] Use human name and english category name for manifest

Commit:
[IMP] multi line notes, unprefix names, pep8 py files

Commit:
[IMP] multi line notes, unprefix names, pep8 py files

Commit:
[IMP] whitespace cleanup in templates

Commit:
[IMP] whitespace cleanup in templates

Commit:
[FIX] base models get _name, custom get _inherit

Commit:
[FIX] base models get _name, custom get _inherit

Commit:
flake fix

Commit:
flake fix

Commit:
add missing spaces

Commit:
add missing spaces

Commit:
prevent crash on menus with no action

Commit:
prevent crash on menus with no action

Commit:
switch GPL-2 to LGPL-3

Commit:
switch GPL-2 to LGPL-3

Commit:
generate license headers with license name

Commit:
generate license headers with license name

Commit:
fix pep/flake errors in generated files

Commit:
fix pep/flake errors in generated files

Commit:
add m2m and relation field options in fields

Commit:
add m2m and relation field options in fields

Commit:
[IMP] Add license field in the view

Commit:
[IMP] Add license field in the view

Commit:
[FIX] Typos

Commit:
[FIX] Typos

Commit:
[IMP] application = True in module manifest

Commit:
[IMP] application = True in module manifest

Commit:
empty selection in m2m field template

Commit:
empty selection in m2m field template

Commit:
[FIX] other licenses, return lines as well

Commit:
[FIX] other licenses, return lines as well

Commit:
[FIX] license not shown in __oe__

Commit:
[FIX] license not shown in __oe__

Commit:
[FIX] unprefix more names, try to get _name/_inherit right [IMP] group by module in zip

Commit:
[FIX] unprefix more names, try to get _name/_inherit right [IMP] group by module in zip

Commit:
[FIX] fix category and summary being on same line

Commit:
[FIX] fix category and summary being on same line

Commit:
[FIX] fix export test

Commit:
[FIX] fix export test

Commit:
[IMP] add tabs for reports/security/workflow/data + partial data/demo generation

Commit:
[IMP] add tabs for reports/security/workflow/data + partial data/demo generation

Commit:
unprefix model names in __init__

Commit:
unprefix model names in __init__

Commit:
[FIX] fix data file names in __openerp__.py

Commit:
[FIX] fix data file names in __openerp__.py

Commit:
[IMP] move Data&Demo after Interface in view

Commit:
[IMP] move Data&Demo after Interface in view

Commit:
[FIX] unprefix view file names

Commit:
[FIX] unprefix view file names

Commit:
[IMP] remove prefixes from field attrs in views

Commit:
[IMP] remove prefixes from field attrs in views

Commit:
[FIX] encode files in zip to utf-8, remove trailing comma in menu groups

Commit:
[FIX] encode files in zip to utf-8, remove trailing comma in menu groups

Commit:
remove unused variable in tests

Commit:
remove unused variable in tests

Commit:
remove AGPL3 or later from license choices: not in base module choices

Commit:
remove AGPL3 or later from license choices: not in base module choices

Commit:
[FIX] Default license

Commit:
[FIX] Default license

Commit:
[IMP] Add OCA as an author

Commit:
[IMP] Add OCA as an author

Commit:
[IMP] Add Bug Tracker section in description

Commit:
[IMP] Add Bug Tracker section in description

Commit:
[IMP] Update description template

Commit:
[IMP] Update description template

Commit:
Merge branch '8.0-prototype' of github.com:savoirfairelinux/server-tools into 8.0-prototype

Commit:
[FIX] Typo in translation

Commit:
[FIX] Allow 2 digits for the minor number

Commit:
wrap default description to pass lint check

Commit:
mis-wrapped url, fix

blaggacao committed:
Update model_name.xml.template

blaggacao committed:
imp(module_prototyper): Add Data Exoprt Feature

blaggacao committed:
Update module_prototyper.py

blaggacao commented:
@osguard can you check / comment on the idea about the ordering problem? can we go into detail here?

blaggacao commented:
we'll post an update tomorrow during the day with some code, stay tuned ;)

OSguard commented:
i m interested if your solution might work. Do our analysis you need a proper graph algorithmen on the structure-data-graph to solve the problem.

blaggacao commented:
As for now, it seems to work, and it's even considerably fast ;)
However, I found this: https://en.wikipedia.org/wiki/Topological_sorting#Algorithms
which is already kind of implemented in `tools/misc.py def topological_sort()`

blaggacao commented:
We are currently stuck on how to trigger the construction of external ids to make references work, before they have first been exported. Any quick ideas? Probably, the csv export method has something reusable built in?

blaggacao commented:
As for now, we don't know how to deal with the problem, that one record can have various external ids, which are return "randomly"; see `get_external_id` in `models.py`. We assume, that this is for a reason, we want to understand better. Or should we just assume, that having more than one external id is not a valid use case and put some (possibly slow) glue code, which makes a clean-up?

blaggacao committed:
add logic for magic

blaggacao committed:
add logic for magic 2

blaggacao committed:
add logic for magic 2 view

blaggacao committed:
super magic final xml

blaggacao commented:
For, now, we could solve most problems
- [x] sorting order of records with parent - child references @OSguard 
- [x] cleaning of magic columns
- [x] creating external IDs in the form `{model_name}_{protoype_technical_name}_{number_with_leading_zeros}`
- [x] autoreferencing of `ref=` attribues by external IDs - this might need revision for many2many @veloutin 
- [x] option to preserve external id, that are present in the database (for data updating)
- [x] sort fields by alphabetic order
- [x] respect model dependencies in file ordering, excluding one2many and cyclical relationships

If decinding to merge, please include (in the correct format) to the contributors (you might need the CLA from jpa@devco.co, still):
(DevCO Colombia is not a legal entity, it's a brand, so legally there are two natural persons committing)

> Juan Pablo Arias (@kurkop), DevCO Colombia, jpa@devco.co, www.devco.co
> David Arnold (@blaggacao), DevCO Colombia, dar@devco.co, www.devco.co

@max3903 @veloutin : may I ask for (gentle) code review? :smile_cat:

OSguard commented:
@ntodorova can you review and test this please

blaggacao commented:
here is a gist for somewhat tedious parent id referencing in one single xml, if you made an export previous to this development...
https://gist.github.com/blaggacao/a56f88588a03ee51bc38

ntodorova commented on module_prototyper/models/module_prototyper.py:
```diff
@@ -503,6 +516,95 @@ def fixup_arch(cls, archstr):
+                # If not unicode (contains ids)
+                # the is not image is hacky and only supports te case when an image is loaded into an "image" field
+                if field != "image" and not isinstance(r[field], (int,float,unicode,str)):
+                    ref = ",".join(r[field].get_external_id().values())
```
I experienced a problem while testing.
I wanted to export some data from hr.employee module.
But if the Employee does not have an img the export is not possible.
The problem is that field 'image_medium' and 'image_small' doesn't exist, so line 591 from module_prototyper.py through an error.

server-tools/module_prototyper/models/module_prototyper.py", line 591, in fixup_data
    ref = ",".join(r[field].get_external_id().values())
AttributeError: 'NoneType' object has no attribute 'get_external_id'

blaggacao committed:
export of images should now work

hack identification of the type of a field by a `isinstance()` negation logic has been replaced by a proper positive identification

blaggacao commented:
@ntodorova can you check again? have a look at new line 590 first.

blaggacao committed:
prevent error if no parent_id exists on the record

the sorting algorithm didn't contemplate the case, where there is no field parent_id. nobrainer.

blaggacao committed:
[fix] parent_id filure and external_id naming

parent_id did not work as intended.
external_id had to be reorderd in order to parallel and reflect the odoo standard import logic.

blaggacao committed:
start support for data FILE order 1

blaggacao committed:
start support for data FILE order 2

blaggacao committed:
on the fastline to file ordering

blaggacao committed:
debuggin 1

Commit:
refatoring toposort WIP

Commit:
small changes WIP

kurkop committed:
ref(module_prototyper/models/module_prototyper.py): Added support to dependencies in generate_data_files

kurkop committed:
ref(module_prototyper/models/module_prototyper.py): Added support to dependencies in generate_data_files

ntodorova commented on module_prototyper/models/module_prototyper.py:
```diff
@@ -433,14 +478,30 @@ def generate_data_files(self):
+                    if records._fields[f].comodel_name != None:
+                        dependencies.append(records._fields[f].comodel_name)
+                # ensure unique values of comodel_types
+            items.append( [model_name, dependencies, records] )
```
While testing I get an error If I have no data - 
server-tools/module_prototyper/models/module_prototyper.py", line 494, in generate_data_files
items.append( [model_name, dependencies, records] )
local variable 'model_name' referenced before assignment

blaggacao committed:
Merge pull request #1 from kurkop/8.0-prototype

8.0 prototype

blaggacao commented:
we basicallz have finished the work, oncorporated toposorts where necesarz and provided a wiyard to tresolv circular dependencies manuallz, tha could not have been resolved automaticallz...final testing is pending, then we'll do an updae commit... stay tuned, i think the outcome is nice ;)

moylop260 commented:
This is a great module.
Thanks for your effort

Small comments:
- 187 commits to add a module is too much, Can you reduce it?
- Can you fix travis fails?

moylop260 commented:
oh sorry
This is a PR to your personal project (no oca) sorry.

Commit:
Merge branch '8.0-prototype' of http://github.com/blaggacao/server-tools into 8.0-prototype

blaggacao commented:
This is closed in favor of a clean #3 , please consider helping to get it merge-ready quickly.
